text/3371-cargo-target-base-dir.md

```diff
@@ -0,0 +1,402 @@
+-->
+
+- Do we want to add forward links for untemplated `CARGO_TARGET_DIR` too in the process of this RFC ?
+- Do we want to differentiate according to users ? `bazel` is a generic build tool, whereas `cargo` is not, so maybe differentiating on users is not necessary for the latter ?
+
+# Future possibilities
+[future-possibilities]: #future-possibilities
+
+<!--
+Think about what the natural extension and evolution of your proposal would
+be and how it would affect the language and project as a whole in a holistic
+way. Try to use this section as a tool to more fully consider all possible
+interactions with the project and language in your proposal.
+Also consider how this all fits into the roadmap for the project
+and of the relevant sub-team.
+
+This is also a good place to "dump ideas", if they are out of scope for the
+RFC you are writing but otherwise related.
+
+If you have tried and cannot think of any future possibilities,
+you may simply state that you cannot think of anything.
+
+Note that having something written down in the future-possibilities section
+is not a reason to accept the current or a future RFC; such notes should be
+in the section on motivation or rationale in this or subsequent RFCs.
+The section merely provides additional information.
+-->
+
+- Introduce remapping into the concept in some way.
+- Introduce a form of garbage collection. Expanded upon this [Zulip thread](https://rust-lang.zulipchat.com/#narrow/stream/246057-t-cargo/topic/last-use.20tracking).
+
+## Adding `XDG_CACHE_HOME`, `HOME` or `CARGO_HOME` as templates
+
+Not all OSes use the XDG convention, notably Windows and macOS (though the latter can be somewhat made to) and it is very easy to define `CARGO_TARGET_DIR=${XDG_CACHE_HOME:-~/.cache}/cargo-target-directories/{manifest-path-hash}` if wanted by users.
+
+It won't work in the `config.toml` but it will work with the environment variable and the command line option, both of which override the TOML config.
+
+It is certainly possible to add at least `{home}` and `{cargo_home}` but it can be done in the future without interfering at all with `{manifest-path-hash}`, making it relatively easy to defer and add if there is demand.
+
+## Use templated `CARGO_TARGET_DIR` as the default instead of `target`
+
+This option has several complications I'm not sure how to resolve:
+
+1. How do we decide on good platform defaults ?
+  - Subsequently, when platform defaults are decided, how do we ensure a new platform has a good default too ?
+  - `CARGO_HOME` is already criticized for being both a *cache* and a *config* home (using the XDG spec semantics), adding more local cache to it in the form of `CARGO_HOME/target-base-dir/` would not improve the situation and should probably not be done, but, if no good alternatives are found, there is precedent to use it for this.
```

Comment:
These look to be more of problems being solved by rust-lang/cargo#1734

Reply:
More my point is that this is a conversation separate from this RFC and should be removed, maybe with a foot note saying any concerns about this should see that issue.

Reply:
Done in e2b877084067d106b0237aa7e9fd3ba210bc4a68